tesseract_collision/include/tesseract_collision/core/collision_margin_data.h

```diff
@@ -0,0 +1,227 @@
+  }
+
+  /**
+   * @brief Get the pairs collision margin data
+   *
+   * If a collision margin for the request pair does not exist it returns the default collision margin data.
+   *
+   * @param obj1 The first object name
+   * @param obj2 The second object name
+   * @return A Vector2d[Contact Distance Threshold, Coefficient]
+   */
+  double getPairCollisionMarginData(const std::string& obj1, const std::string& obj2) const
+  {
+    auto key = tesseract_common::makeOrderedLinkPair(obj1, obj2);
+    const auto it = lookup_table_.find(key);
+
+    if (it != lookup_table_.end())
+      return it->second;
+
+    return default_collision_margin_;
+  }
+
+  /**
+   * @brief Get Collision Margin Data for stored pairs
+   * @return A map of link pairs collision margin data
+   */
+  const PairsCollisionMarginData& getPairsCollisionMarginData() const { return lookup_table_; }
+
+  /**
+   * @brief Get the largest collision margin
+   *
+   * This used when setting the contact distance in the contact manager.
+   *
+   * @return Max contact distance threshold
+   */
+  double getMaxCollisionMargin() const { return max_collision_margin_; }
+
+  /**
+   * @brief Increment all margins by input amount. Useful for inflating or reducing margins
+   * @param increment Amount to increment margins
+   */
+  void incrementMargins(const double& increment)
+  {
+    default_collision_margin_ += increment;
+    max_collision_margin_ += increment;
+    for (auto& pair : lookup_table_)
+      pair.second += increment;
+  }
+
+  /**
+   * @brief Scale all margins by input value
+   * @param scale Value by which all margins are multipled
+   */
+  void scaleMargins(const double& scale)
+  {
+    default_collision_margin_ *= scale;
+    max_collision_margin_ *= scale;
+    for (auto& pair : lookup_table_)
+      pair.second *= scale;
+  }
+
+  /**
+   * @brief Apply the contents of the provide CollisionMarginData based on the override type
+   * @param collision_margin_data The collision margin data to apply
+   * @param override_type The type indicating how the provided data should be applied.
+   */
+  void apply(const CollisionMarginData& collision_margin_data, CollisionMarginOverrideType override_type)
+  {
+    switch (override_type)
+    {
+      case CollisionMarginOverrideType::REPLACE:
+      {
+        *this = collision_margin_data;
+        break;
+      }
+      case CollisionMarginOverrideType::OVERRIDE_DEFAULT_MARGIN:
+      {
+        default_collision_margin_ = collision_margin_data.default_collision_margin_;
+        updateMaxCollisionMargin();
+        break;
+      }
+      case CollisionMarginOverrideType::OVERRIDE_PAIR_MARGIN:
+      {
+        lookup_table_ = collision_margin_data.lookup_table_;
+        updateMaxCollisionMargin();
+        break;
+      }
+      case CollisionMarginOverrideType::MODIFY_PAIR_MARGIN:
+      {
+        lookup_table_.insert(collision_margin_data.lookup_table_.begin(), collision_margin_data.lookup_table_.end());
```

Comment:
Perhaps we should check the return value here and print a debug message if any of them failed.

Reply:
I do not believe this should ever fail; should it?

Reply:
Ok, I think I understand what you are referring to. The insert approach will not update, it will only add those which do not exist which was not what I thought it did. I am updating it such that it will update and add those that do not exist.

Reply:
Yeah this isn't relevant since you changed it to update pairs that already existed. I like that better anyway.